visualization/generate_experimentId.py

```diff
@@ -0,0 +1,86 @@
+import pandas as pd
+import numpy as np
+import numbers
+
+
+def generate_experimentId(measurement_data):
+    '''
+    automatically generate the experimentId:
+    every row which shares the observableParameters (if provided),
+    noiseParameters and observableTransformation gets
+    the same experimentId, for the assignment all numeric values in the
+    noiseParameters are considered to be the same
+
+    Parameters:
+    ----------
+
+    measurement_data: pandas data frame for measurement data
+
+    Return:
+    ----------
+
+    measurement_data: pandas data frame for measurement data with column for
+    experimentId added
+
+    '''
+    observableParameters = np.array(measurement_data.observableParameters)
+    noiseParameters = np.array(measurement_data.noiseParameters)
+    observableTransformation = np.array(
+        measurement_data.observableTransformation
+    )
+
+    # check if there occurs the name 'nan' in the observableParameter column
+    # , should not occur!
+    checkNaming = np.where(observableParameters == 'empty')
+    if checkNaming[0].size > 0:
+        print('error')
+
+    # but noiseParameters should always has an entry
+    checkNaming = np.where(noiseParameters == 'empty')
+    if checkNaming[0].size > 0:
+        print('error')
+
+    # help functions to check for numeric values and nans in arrays
+    isnumeric = np.vectorize(isinstance)
+    isnanVectorized = np.vectorize(np.isnan)
+
+    # assign 'empty' to missing observableParameters
+    tmp_observableParameters = observableParameters
+    tmp_ind = np.where(isnumeric(observableParameters, numbers.Number))
+    tmp_observableParameters[tmp_ind[0][
+        np.where(isnanVectorized(observableParameters[tmp_ind[0]]))]
+    ] = 'empty'
```

Comment:
It would be good it the used measurement_data df is not modified in the function, to avoid problems with later calls on it. I think that the methods used here are not in-place, so there shouldn't be a problem. Otherwise, a measurement_df = copy.deepcopy(measurement_df) at the beginning of the file might help.